### PR TITLE
Model serve

### PR DESCRIPTION
**Pull request by dizcology**
A simple configuration and step-by-step guide for setting up GAE and Cloud Endpoint to serve a pickled sklearn model saved on GCS.

### CODE REVIEW

File: gae_serve/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    - Replace `PROJECT_ID` with your project's id.
+
+    - Replace `BUCKET_NAME` with the name of the bucket you created on GCS above.
+
+    - Replace `CONFIG_ID` with the config id you got from the service endpoint deployment.
+
+1. Deploy the backend service:
+
+    `gcloud app deploy`
```
[the remark this answers is not in the document]
Comment (nkashy1) [reply gae_serve/README.md]: We should warn users about the long deploy times.

File: gae_serve/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+
+1. If the deployment is successful, you can test it from the command line.  Firt go to your project's console's IAM page and create an API key `API_KEY`, and use it and your project's id in the following:
```
Comment (nkashy1) [inline gae_serve/README.md]: The API key addition should be fleshed out - either using links to Pantheon docs or by adding screenshots here.

File: gae_serve/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    `curl -H "Content-Type: application/json" -X POST -d '{"X": [[1, 2], [5, -1], [1, 0]]}' "https://modelserve-dot-PROJECT_ID.appspot.com/predict?key=API_KEY"`
+
+    You should get the following response:
+
+    `{"y": [0.6473534912754967, -0.7187842827829021, 0.3882338314071392]}`
```
Comment (nkashy1) [inline gae_serve/README.md]: I got the following response:

```
(gae-serve)nkash@nkash:~/GoogleCloudPlatform/ml-on-gcp/gae_serve$ curl -H "Content-Type: application/json" -X POST -d '{"X": [[1, 2], [5, -1], [1, 0]]}' "https://modelserve-dot-PROJECT_ID.appspot.com/predict?key=$GAE_SERVE_API_KEY"
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Page not found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL was not found on this server.  <ins>That’s all we know.</ins>
```

Because of firewall, though, probably.

Comment (nkashy1) [reply gae_serve/README.md]: Actually, the reason I was getting that error was because I forgot to replace the `PROJECT_ID` in the URL in my curl command. Works now.

File: gae_serve/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    Note that this file defines the API specifying the input `X` to be an array of arrays of floats, and output `y` to be an array of floats.  The model included in this sample code is a linear regression model with 2-dimensional inputs.
+
+1. Deploy the service endpoint:
+
+    `gcloud service-management deploy modelserve.yaml`
+
+    This step deploys a [Cloud Endpoint](https://cloud.google.com/endpoints/) service, which allows us to monitor the API usage.
+
+1. If the deploymenti s successful, get the deployment's config id from console.  It should look like `2017-08-03r0`.
```
Comment (p-buse) [inline gae_serve/README.md]: Typo: "deploymenti s"

File: gae_serve/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+1. `cd ml-on-gcp`
+
+1. `pip install -r requirements.txt`
+
+1. The name `modelserve` is tentative which can be changed to something else, but you need to change all occurrences consistently.
+
+1. Update `modelserve.yaml`:  Replace `PROJECT_ID` with your project's id in this line:
+
+    `host: "modelserve.endpoints.PROJECT_ID.cloud.goog"`
```
Comment (p-buse) [inline gae_serve/README.md]: Because this example runs on GAE, this host should have the form `PROJECT_ID.appspot.com`.

File: gae_serve/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    Note that this file defines the API specifying the input `X` to be an array of arrays of floats, and output `y` to be an array of floats.  The model included in this sample code is a linear regression model with 2-dimensional inputs.
```
Comment (p-buse) [inline gae_serve/README.md]: Nit: why use uppercase 'X' and lowercase 'y'?

File: gae_serve/README.md (hunk 4 above)
Comment (p-buse) [inline gae_serve/README.md]: It would be helpful to mention that if the user misses the Config ID the first time, they can type
`gcloud service-management configs list --service="PROJECT_ID.appspot.com"` to show all deployed configs.

File: gae_serve/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+1. Create a GCS bucket and copy the model file over:
+
+    ```
+    gsutil mb gs://BUCKET_NAME
+    gsutil cp lr.pkl gs://BUCKET_NAME
+    ```
+
+1. Update `app.yaml`:  
+
+    - Replace `PROJECT_ID` with your project's id.
```
Comment (p-buse) [inline gae_serve/README.md]: Please clarify that this is a GCP project ID.

File: gae_serve/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    `curl -H "Content-Type: application/json" -X POST -d '{"X": [[1, 2], [5, -1], [1, 0]]}' "https://modelserve-dot-PROJECT_ID.appspot.com/predict?key=API_KEY"`
+
+    You should get the following response:
+
+    `{"y": [0.6473534912754967, -0.7187842827829021, 0.3882338314071392]}`
+
+
+
+
+
```
Comment (p-buse) [inline gae_serve/README.md]: Please delete unnecessary whitespace.

File: gae_serve/app.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+entrypoint: gunicorn -b :$PORT main:app
+
+runtime_config:
+  python_version: 3
+
+manual_scaling:
+  instances: 1
+
+endpoints_api_service:
+  name: modelserve.endpoints.PROJECT_ID.cloud.goog
```
Comment (p-buse) [inline gae_serve/app.yaml]: name: ... should be `PROJECT_ID.appspot.com`.

File: gae_serve/app.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  config_id: CONFIG_ID
+
+env_variables:
+    # gs://BUCKET_NAME/lr.pkl
```
Comment (p-buse) [inline gae_serve/app.yaml]: Why is there a commented GCS link here?

File: gae_serve/modelserve.yaml [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+---
+swagger: "2.0"
+info:
+  title: "modelserve"
+  description: "Get prediction from sklearn model."
+  version: "1.0.0"
+host: "modelserve.endpoints.PROJECT_ID.cloud.goog"
```
Comment (p-buse) [inline gae_serve/modelserve.yaml]: Change to `PROJECT_ID.appspot.com` here as well.

File: gae_serve/modelserve.yaml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  y:
+    type: "object"
+    required:
+    - "y"
+    properties:
+      y:
+        type: "array"
+        items:
+          type: "number"
+          format: "float"
```
Comment (p-buse) [inline gae_serve/modelserve.yaml]: Add newline here, please.

File: gae_serve/README.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+The benefits of this configuration include:
+
+1. GAE's autoscaling and load balancing.
+1. Cloud Endpoints' monitoring and access control.
+
+
+## Requirements
+
+Python.
```
[the remark this answers is not in the document]
Comment (thedriftofwords) [reply gae_serve/README.md]: When you use an acronym for the first time in a document, say what it is first.

For example: 

"Create a Google App Engine (GAE) service that loads a pickled sci-kit learn model from Google Cloud Storage (GCS), and use it to serve prediction requests through Cloud Endpoints."

Then, throughout the rest of the document it's OK to say "GAE" and "GCS".

File: gae_serve/README.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+GCP SDK.  gsutil.
+
+A GCP project that has GAE, GCS, and Cloud Endpoints enabled.
```
[the remark this answers is not in the document]
Comment (thedriftofwords) [reply gae_serve/README.md]: +1 to comments about more links.

Did you mean for "GCP SDK" and "gsutil" to be on the same line? According to the [gsutil docs](https://cloud.google.com/storage/docs/gsutil#gettingstarted), gsutil is included in the GCP SDK. Could you say more about why you specifically listed gsutil here, and add a link to the docs as well?

File: gae_serve/README.md (hunk 6 above)
Comment (thedriftofwords) [reply gae_serve/README.md]: @p-buse Uppercase 'X' is a common notation for the 2D array of inputs, and similarly, lowercase 'y' typically denotes a 1D array of outputs - throughout the scikit-learn documentation and all the slides in my data science class.

File: gae_serve/README.md (hunk 2 above)
Comment (thedriftofwords) [reply gae_serve/README.md]: Typo: "Firt" --> "First"

File: gae_serve/README.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    - Replace `CONFIG_ID` with the config id you got from the service endpoint deployment.
+
+1. Deploy the backend service:
+
+    `gcloud app deploy`
+
+
+1. If the deployment is successful, you can test it from the command line.  Firt go to your project's console's IAM page and create an API key `API_KEY`, and use it and your project's id in the following:
+
+    `curl -H "Content-Type: application/json" -X POST -d '{"X": [[1, 2], [5, -1], [1, 0]]}' "https://modelserve-dot-PROJECT_ID.appspot.com/predict?key=API_KEY"`
```
[the remark this answers is not in the document]
Comment (thedriftofwords) [reply gae_serve/README.md]: +1 to more narrative here

File: gae_serve/main.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+from google.cloud import storage
+from flask import Flask, request
+
+MODEL_BUCKET = os.environ['MODEL_BUCKET']
+MODEL_FILENAME = os.environ['MODEL_FILENAME']
+MODEL = None
+
+app = Flask(__name__)
+
+# NOTE: the first response could be slower if the model is large.
```
Comment (dizcology) [inline gae_serve/main.py]: This can be an issue - need to experiment what happens if the model takes more time to load.  GAE has a 30 seconds timeout for each request.

One possible approach is to use healthcheck to initiate model loading, and health check should not return OK until the model has been loaded.

File: gae_serve/README.md (hunk 3 above)
Comment (dizcology) [reply gae_serve/README.md]: added note in README to reminder the user.

File: gae_serve/README.md (hunk 8 above)
Comment (dizcology) [reply gae_serve/README.md]: done

File: gae_serve/app.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (dizcology) [reply gae_serve/app.yaml]: indeed `$PORT` is on the instance.

`PROJECT_ID` probably could be set to, if we change the instruction to something like `export PROJECT_ID` but I really hope to force the user to open up the config yaml file.

Comment (dizcology) [reply gae_serve/app.yaml]: done.

File: gae_serve/app.yaml (hunk 10 above)
Comment (dizcology) [reply gae_serve/app.yaml]: trying to point out that the app would be looking for this file at that particular URI.  added to this comment to make it clearer.

File: gae_serve/modelserve.yaml (hunk 11 above)
Comment (dizcology) [reply gae_serve/modelserve.yaml]: done.

File: gae_serve/modelserve.yaml (hunk 12 above)
Comment (dizcology) [reply gae_serve/modelserve.yaml]: done.

File: gae_serve/README.md (hunk 6 above)
Comment (nkashy1) [reply gae_serve/README.md]: TIL

File: gae_serve/main.py (hunk 16 above)
Comment (nkashy1) [reply gae_serve/main.py]: @dizcology : Oh, nice, I like the health check idea. Does the 30 second limit apply to health check?

File: gae_serve/README.md [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+        - Replace `PROJECT_ID` with your GCP project's id.
+
+        - Replace `BUCKET_NAME` with the name of the bucket you created on GCS above.
+
+        - Replace `CONFIG_ID` with the configuration ID you got from the service endpoint deployment.
+
+    * If you have not deployed any service to GAE with this GCP project:
+
+        - Replace `service: modelserve` with `service: default` in `app.yaml`.  The first service deployed to GAE must be named `default`.
```
Comment (nkashy1) [inline gae_serve/README.md]: *Temporarily* replace - can you specify a different YAML file when deploying to GAE? If so, would be good to include both YAML files in the repo.

File: gae_serve/README.md [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+        - Replace `PROJECT_ID` with your GCP project's id.
+
+        - Replace `BUCKET_NAME` with the name of the bucket you created on GCS above.
+
+        - Replace `CONFIG_ID` with the configuration ID you got from the service endpoint deployment.
+
+        - **Note that you will not be able to delete the `default` service from your project.**
```
Comment (nkashy1) [inline gae_serve/README.md]: Would be good to give an explanation of the role of the default service. What purpose does it serve and why is it not delete-able?

Our audience is probably not super familiar with this kind of procedure.

File: gae_serve/README.md [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+1. If the deployment is successful, you can test it from the command line:
+
+    - Create an API key with the "Create credentials" button on the [Credentials](https://console.cloud.google.com/apis/credentials) page.  Make sure you switch to the correct GCP project first.
+
+    - Now you can test the service endpoint: (Remember to replace `PROJECT_ID` and `API_KEY` with their actual values below.)
+
+        * If you have `service: modelserve` in `app.yaml`:
+
+            `curl -H "Content-Type: application/json" -X POST -d '{"X": [[1, 2], [5, -1], [1, 0]]}' "https://modelserve-dot-PROJECT_ID.appspot.com/predict?key=API_KEY"`
```
Comment (nkashy1) [inline gae_serve/README.md]: Could we actually include a little Python client to make this request? Not sure how well this would work for Windows users.

Comment (nkashy1) [reply gae_serve/README.md]: This would in general also be useful as it would give them a template client that they could use to make requests against their service - teaching, for example, how to read `API_KEY` (from environment? from config?) and similarly for `PROJECT_ID`.

File: gae_serve/README.md [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+* If the service was deployed as the `modelserve` service (that is, with `service: modelserve` in `app.yaml`), then you can delete service by running:
+
+    `gcloud app services delete modelserve`
+
+    `gcloud service-management delete PROJECT_ID.appspot.com`
+
+
+## Advanced usage
+
+### Healthcheck
```
Comment (nkashy1) [inline gae_serve/README.md]: Not important for this iteration of work, but would like to eventually give users a guide to using updated health checks for model serving in the way you suggested in previous comments.

File: gae_serve/app.yaml [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+service: modelserve
+runtime: python
+env: flex
+entrypoint: gunicorn -b :$PORT main:app
+
+runtime_config:
+  python_version: 3
+
+manual_scaling:
+  instances: 1
```
[the remark this answers is not in the document]
Comment (nkashy1) [reply gae_serve/app.yaml]: This would be great, let's make an issue for this.

File: gae_serve/README.md (hunk 19 above)
Comment (nkashy1) [reply gae_serve/README.md]: Might be good to ask @SurferJeffAtGoogle to go through this whole workflow once and measure the amount of rage induced.

File: gae_serve/README.md (hunk 17 above)
Comment (dizcology) [reply gae_serve/README.md]: I hope to keep the structure of the sample in close to the same shape as would be for a setup in a real production system, where only two yaml files are needed (backend service and service endpoint), and so I hope to avoid adding an additional yaml and prefer forcing the user to edit the files and learn about the caveats.

File: gae_serve/README.md (hunk 18 above)
Comment (dizcology) [reply gae_serve/README.md]: I haven't been able to find an explanation of why the default app cannot be deleted - seems to be some technical reasons.

File: gae_serve/README.md (hunk 19 above)
Comment (dizcology) [reply gae_serve/README.md]: Adding a client is a great idea - using swagger/OpenAI should make it easy.  I will do this before merge.

File: gae_serve/README.md (hunk 20 above)
Comment (dizcology) [reply gae_serve/README.md]: I moved this into just a link because it seems that even with the default "legacy" health check, the model is loaded upon the first (health check) request, and the service isn't considered ready until that process is finished (the flask app would be blocking while loading the model).  So there's no additional work needed to use health check to trigger model loading.

File: README.md [hunk 22]
```diff
@@ -15,3 +15,11 @@ really care about.
 1. [Estimators](tensorflow/tf-estimators.ipynb) - A guide to the Estimator
    interface.
 
+
+- - - 
+
+## scikit-learn
+
+1. [Model serve](sklearn/gae_serve/README.md) - Serve model with Google App
```
Comment (dizcology) [inline README.md]: I moved this sample under sklearn.  Most definitely the setup itself can be used for models trained with other framework (comment added for this), but I hope to get sklearn users' attention more quickly this way.
